core/src/main/java/nl/nn/adapterframework/util/MessageUtils.java

```diff
@@ -87,16 +78,121 @@ public static String computeContentType(Message message) {
 			return null;
 		}
 
-		String mimeType = (String)message.getContext().get(MessageContext.METADATA_MIMETYPE);
-		if(StringUtils.isEmpty(mimeType)) {
+		MimeType mimeType = (MimeType)message.getContext().get(MessageContext.METADATA_MIMETYPE);
+		if(mimeType == null) {
 			return null;
 		}
 
-		StringBuilder contentType = new StringBuilder(mimeType);
-		if(!message.isBinary()) {
+		StringBuilder contentType = new StringBuilder();
+		contentType.append(mimeType.getType());
+		contentType.append('/');
+		contentType.append(mimeType.getSubtype());
+
+		if(message.getCharset() != null) {
 			contentType.append(";charset=");
 			contentType.append(message.getCharset());
 		}
 		return contentType.toString();
 	}
+
+	/**
+	 * Reads the first 10k bytes of (binary) messages to determine the charset when not present in the {@link MessageContext}.
+	 * @throws IOException when it cannot read the first 10k bytes.
+	 */
+	public static Charset computeDecodingCharset(Message message) throws IOException {
+		return computeDecodingCharset(message, charsetConfidenceLevel);
+	}
+
+	/**
+	 * Reads the first 10k bytes of (binary) messages to determine the charset when not present in the {@link MessageContext}.
+	 * @param confidence percentage required to successfully determine the charset.
+	 * @throws IOException when it cannot read the first 10k bytes.
+	 */
+	public static Charset computeDecodingCharset(Message message, int confidence) throws IOException {
+		if(Message.isEmpty(message) || !message.isBinary()) {
+			return null;
+		}
+
+		if(StringUtils.isNotEmpty(message.getCharset()) && !StreamUtil.AUTO_DETECT_CHARSET.equalsIgnoreCase(message.getCharset())) {
+			return Charset.forName(message.getCharset());
+		}
+
+		CharsetDetector detector = new CharsetDetector();
+		detector.setText(message.getMagic());
+		CharsetMatch match = detector.detect();
+		String charset = match.getName();
+
+		if(match.getConfidence() > 90) {
+			LOG.debug("update charset for message [{}], full match [{}] with confidence level [{}/{}]", message, charset, match.getConfidence(), confidence);
+			return updateMessageCharset(message, charset);
+		}
+
+		//Guesstimate, encoding is not UTF-8 but either CP1252/Latin1/ISO-8859-1.
+		if(charset.startsWith("windows-125")) {
+			charset = "windows-1252";//1250/1/3 have a combined adoption rate of 1.6% assume 1252 instead!
+		}
+		if(match.getConfidence() >= confidence) {
+			LOG.debug("update charset for message [{}], potential match [{}] with confidence level [{}/{}]", message, charset, match.getConfidence(), confidence);
+			return updateMessageCharset(message, charset);
+		}
+
+		LOG.info("unable to detect charset for message [{}] closest match [{}] did not meet confidence level [{}/{}]", message, charset, match.getConfidence(), confidence);
+		return updateMessageCharset(message, null); //return NULL so calling method can fall back to the default charset.
+	}
+
```

Comment:
Ik denk dat je hier ook ergens wil regelen dat als er op een repeatable message meerdere malen getReader(auto) gedaan wordt, er dan maar één keer gedetecteerd wordt, ook als dat niet succesvol is.